### PR TITLE
v15/resource/encryptionkey: simplify resource

### PR DESCRIPTION
**Pull request by kopiczko**
https://github.com/giantswarm/giantswarm/issues/3783

I'm getting my hands dirty. I want to get rid of this ugly thing in cloudformation resource: https://github.com/giantswarm/aws-operator/blob/dc8780cd58f69bf1bc9be241767a4a4f3315085f/service/controller/v15/resource/cloudformation/create.go#L32-L42

I barely could understand what is going on while this resource does very simple thing.

Next steps is to make sure `addRoleAccess` is just reconciled outside cloudformation resource. This feels hacky and unnecessary.

### CODE REVIEW

File: service/controller/v15/resource/encryptionkey/delete.go [hunk 1]
```diff
@@ -4,60 +4,33 @@ import (
 	"context"
 
 	"github.com/giantswarm/microerror"
-	"github.com/giantswarm/operatorkit/controller"
 
-	"github.com/giantswarm/aws-operator/service/controller/v15/encrypter"
+	"github.com/giantswarm/aws-operator/service/controller/v15/key"
 )
```
Comment (kopiczko) [inline service/controller/v15/resource/encryptionkey/delete.go]: The diff is ugly. File content is:

```go
package encryptionkey

import (
	"context"

	"github.com/giantswarm/microerror"

	"github.com/giantswarm/aws-operator/service/controller/v15/key"
)

func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
	customObject, err := key.ToCustomObject(obj)
	if err != nil {
		return microerror.Mask(err)
	}

	{
		r.logger.LogCtx(ctx, "level", "debug", "message", "ensuring deletion of encryption key")

		current, err := r.encrypter.CurrentState(ctx, customObject)
		if err != nil {
			return microerror.Mask(err)
		}

		if current.KeyName != "" {
			err = r.encrypter.DeleteKey(ctx, current.KeyName)
			if err != nil {
				return microerror.Mask(err)
			}
		}

		r.logger.LogCtx(ctx, "level", "debug", "message", "ensured deletion of encryption key")
	}

	return nil
}
```

File: service/controller/v15/resource/encryptionkey/create.go [hunk 2]
```diff
@@ -8,45 +8,34 @@ import (
 	"github.com/giantswarm/aws-operator/service/controller/v15/key"
 )
```
Comment (kopiczko) [inline service/controller/v15/resource/encryptionkey/create.go]: The diff is ugly. File content is:

```go
package encryptionkey

import (
	"context"

	"github.com/giantswarm/microerror"

	"github.com/giantswarm/aws-operator/service/controller/v15/key"
)

func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
	customObject, err := key.ToCustomObject(obj)
	if err != nil {
		return microerror.Mask(err)
	}

	{
		r.logger.LogCtx(ctx, "level", "debug", "message", "ensuring encryption key")

		current, err := r.encrypter.CurrentState(ctx, customObject)
		if err != nil {
			return microerror.Mask(err)
		}

		desired, err := r.encrypter.DesiredState(ctx, customObject)
		if err != nil {
			return microerror.Mask(err)
		}

		if current.KeyName == "" {
			err = r.encrypter.CreateKey(ctx, customObject, desired.KeyName)
			if err != nil {
				return microerror.Mask(err)
			}
		}

		r.logger.LogCtx(ctx, "level", "debug", "message", "ensured encryption key")
	}

	return nil
}
```

File: service/controller/v15/resource/encryptionkey/create.go [hunk 3]
```diff
@@ -8,45 +8,34 @@ import (
...
+		if current.KeyName == "" {
+			err = r.encrypter.CreateKey(ctx, customObject, desired.KeyName)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		}
 
-	if currentKeyState.KeyName == "" || desiredKeyState.KeyName != currentKeyState.KeyName {
-		return desiredKeyState, nil
+		r.logger.LogCtx(ctx, "level", "debug", "message", "ensured encryption key")
```
Comment (xh3b4sd) [inline service/controller/v15/resource/encryptionkey/create.go]: Do we go now with the "ensure" message for logs here or is it the "creating" messaging? I feel like there are some differences in what we express and which information we provide. The latter option would also give information on if the creation was done in this very reconciliation loop or not. With the former we only know it is a done deal but we do not know when and how and kind of it at all.

File: service/controller/v15/resource/encryptionkey/resource.go [hunk 4]
```diff
@@ -8,81 +8,45 @@ import (
 )
 
 const (
-	// Name is the identifier of the resource.
-	Name = "kmskeyv14"
+	name = "kmskeyv14"
```
Comment (xh3b4sd) [inline service/controller/v15/resource/encryptionkey/resource.go]: That doesn't look right BTW.

File: service/controller/v15/resource/encryptionkey/resource.go [hunk 5]
```diff
@@ -8,81 +8,45 @@ import (
...
 )
 
-// Config represents the configuration used to create a new cloudformation resource.
 type Config struct {
-	// Dependencies.
 	Encrypter encrypter.Interface
 	Logger    micrologger.Logger
 
-	// Settings.
 	InstallationName string
```
Comment (xh3b4sd) [inline service/controller/v15/resource/encryptionkey/resource.go]: Is the installation name used for anything?

File: service/controller/v15/resource/encryptionkey/mock.go [hunk 6]
```diff
@@ -1,50 +0,0 @@
...
-		return encrypter.EncryptionKeyState{}, fmt.Errorf("could not get current state")
-	}
-
-	return encrypter.EncryptionKeyState{
-		KeyID:   e.keyID,
-		KeyName: e.keyName,
-	}, nil
-}
-
-func (e *EncrypterMock) EncryptionKey(ctx context.Context, customObject v1alpha1.AWSConfig) (string, error) {
```
Comment (xh3b4sd) [inline service/controller/v15/resource/encryptionkey/mock.go]: What happened to these mocks? Don't we need them? Why have they been there at all in the first place?

Comment (kopiczko) [reply service/controller/v15/resource/encryptionkey/mock.go]: I think they were used for tests. Tests are gone so this is gone.

File: service/controller/v15/resource/encryptionkey/resource.go (hunk 4 above)
Comment (kopiczko) [reply service/controller/v15/resource/encryptionkey/resource.go]: What is wrong?

Comment (xh3b4sd) [reply service/controller/v15/resource/encryptionkey/resource.go]:
```
service/controller/v15/resource/encryptionkey/resource.go
```

```
kmskeyv14
```

File: service/controller/v15/resource/encryptionkey/resource.go (hunk 5 above)
Comment (kopiczko) [reply service/controller/v15/resource/encryptionkey/resource.go]: No. Good catch. Removed.

File: service/controller/v15/resource/encryptionkey/resource.go (hunk 4 above)
Comment (xh3b4sd) [reply service/controller/v15/resource/encryptionkey/resource.go]: Should rather be `encryptionkeyv15` IMO.

Comment (kopiczko) [reply service/controller/v15/resource/encryptionkey/resource.go]: LoL, eagle eye.

Comment (kopiczko) [reply service/controller/v15/resource/encryptionkey/resource.go]: Fixed.

File: service/controller/v15/resource/encryptionkey/create.go (hunk 3 above)
Comment (kopiczko) [reply service/controller/v15/resource/encryptionkey/create.go]: I tried to add something here https://github.com/giantswarm/aws-operator/pull/1096/commits/68823c6342a5bc65355021bcadfedb8d0d9448be

The problem is this resource doesn't gather facts. I will try to improve that further, as this should what `addRoleAccess` does currently (see description).

**Pull request merged**